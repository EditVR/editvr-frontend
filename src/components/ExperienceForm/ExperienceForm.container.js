/**
 * @file ExperienceForm.container.js
 * Exports a redux-connected ExperienceForm component.
 */

import { connect } from 'react-redux';

import ExperienceForm from './ExperienceForm';

const mapDispatchToProps = dispatch => ({
  dispatch
});

const mapState = ({ user, experiences }) => ({
  user,
  experiences
});

export default connect(mapState, mapDispatchToProps)(ExperienceForm);
