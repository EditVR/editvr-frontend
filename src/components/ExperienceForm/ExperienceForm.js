/**
 * @file ExperienceForm.js
 * Exports a component that allows users to operate on experiences.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TextField, Button, withStyles } from '@material-ui/core';

import { Message } from '../';
import ExperienceFormStyles from './ExperienceForm.style';
import { EXPERIENCES_CREATE } from '../../constants';

class ExperienceForm extends Component {
  static propTypes = {
    submitHandler: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
    experienceSlug: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    classes: PropTypes.shape({
      textField: PropTypes.string.isRequired,
      button: PropTypes.string.isRequired
    }).isRequired,
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.shape({
      push: PropTypes.func.isRequired
    }).isRequired,
    user: PropTypes.shape({
      authentication: PropTypes.shape({
        accessToken: PropTypes.string.isRequired,
        csrfToken: PropTypes.string.isRequired
      }).isRequired
    }).isRequired,
    experiences: PropTypes.shape({
      error: PropTypes.string,
      items: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string.isRequired,
          title: PropTypes.string.isRequired,
          body: PropTypes.PropTypes.shape({
            value: PropTypes.string
          }),
          field_experience_path: PropTypes.string.isRequired
        })
      )
    })
  };

  static defaultProps = {
    submitHandler: false,
    experienceSlug: false,
    experiences: {
      error: null,
      items: []
    }
  };

  /**
   * Handles a creating an experience.
   *
   * @param {object} event - Submit event object.
   */
  handleCreate = event => {
    event.preventDefault();
    const {
      dispatch,
      user,
      history: { push }
    } = this.props;
    dispatch({
      type: EXPERIENCES_CREATE,
      user,
      title: event.target[0].value,
      field_experience_path: event.target[1].value,
      body: event.target[2].value,
      successHandler: () => push('/dashboard')
    });
  };

  /**
   * {@inheretdoc}
   */
  render() {
    const {
      classes,
      submitHandler,
      experienceSlug,
      experiences: { error, items }
    } = this.props;

    // If this is an editorial form, grab the existing experience.
    const experience = experienceSlug
      ? items.find(item => item.field_experience_path === experienceSlug)
      : false;

    return (
      <form onSubmit={submitHandler || this.handleCreate}>
        {error && <Message>{error}</Message>}
        <TextField
          id="title"
          label="Title"
          type="text"
          required
          helperText="Enter a user-friendly title for your experience."
          defaultValue={experience ? experience.title : ''}
          className={classes.textField}
        />
        <TextField
          id="field_experience_path"
          label="URL Path"
          type="text"
          required
          helperText="Enter a name for your experience. For example, if you enter 'my-new-experience', your experience will be published to /experience/my-new-experience."
          defaultValue={experience ? experience.field_experience_path : ''}
          className={classes.textField}
        />
        <TextField
          id="body"
          label="Description"
          type="text"
          multiline
          required
          rows={6}
          helperText="Describe your experience so people will know what to expect before they enter your experience."
          defaultValue={
            experience && experience.body ? experience.body.value : ''
          }
          className={classes.textField}
        />
        <Button
          variant="raised"
          color="primary"
          type="submit"
          className={classes.button}
        >
          Create
        </Button>
      </form>
    );
  }
}

export default withStyles(ExperienceFormStyles)(ExperienceForm);
