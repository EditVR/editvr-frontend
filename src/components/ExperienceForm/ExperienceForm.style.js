/**
 * @file ExperienceCreateForm.style.js
 * Exports ExperienceCreateForm component styles.
 */

export default theme => ({
  textField: {
    width: '100%',
    marginTop: `${theme.spacing.unit * 2}px`
  },
  button: {
    marginTop: `${theme.spacing.unit * 3}px`,
    marginRight: theme.spacing.unit
  }
});
