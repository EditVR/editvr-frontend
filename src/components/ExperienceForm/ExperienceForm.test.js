/**
 * @file ExperienceForm.test.js
 * Contains tests for ExperienceForm.js.
 */

import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import ExperienceForm from './ExperienceForm.container';

describe('<ExperienceForm />', () => {
  it('Matches its snapshot', () => {
    const store = configureStore()({
      user: {
        error: null,
        authentication: {
          accessToken: 'test',
          csrfToken: 'test'
        }
      }
    });

    expect(
      renderer
        .create(
          <Router>
            <ExperienceForm store={store} />
          </Router>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
