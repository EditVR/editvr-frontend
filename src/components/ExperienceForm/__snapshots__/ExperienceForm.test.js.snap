// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExperienceForm /> Can be a Create form for Experiences 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    aria-describedby="title-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="title"
    >
      Title
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25"
        defaultValue=""
        disabled={false}
        id="title"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="title-helper-text"
    >
      Enter a user-friendly title for your experience.
    </p>
  </div>
  <div
    aria-describedby="field_experience_path-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="field_experience_path"
    >
      URL Path
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25"
        defaultValue=""
        disabled={false}
        id="field_experience_path"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="field_experience_path-helper-text"
    >
      Enter a name for your experience. For example, if you enter 'my-new-experience', your experience will be published to /experience/my-new-experience.
    </p>
  </div>
  <div
    aria-describedby="body-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="body"
    >
      Description
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-multiline-23 MuiInput-underline-21"
    >
      <textarea
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25 MuiInput-inputMultiline-27"
        defaultValue=""
        disabled={false}
        id="body"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={6}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="body-helper-text"
    >
      Describe your experience so people will know what to expect before they enter your experience.
    </p>
  </div>
  <button
    className="MuiButtonBase-root-50 MuiButton-root-34 MuiButton-raised-40 MuiButton-raisedPrimary-41 ExperienceForm-button-2"
    disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="submit"
  >
    <span
      className="MuiButton-label-35"
    >
      Save
    </span>
    <span
      className="MuiTouchRipple-root-53"
    />
  </button>
</form>
`;

exports[`<ExperienceForm /> Can be an Edit form for Experiences 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    aria-describedby="title-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="title"
    >
      Title
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25"
        defaultValue="test"
        disabled={false}
        id="title"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="title-helper-text"
    >
      Enter a user-friendly title for your experience.
    </p>
  </div>
  <div
    aria-describedby="field_experience_path-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="field_experience_path"
    >
      URL Path
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25"
        defaultValue="test"
        disabled={false}
        id="field_experience_path"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="field_experience_path-helper-text"
    >
      Enter a name for your experience. For example, if you enter 'my-new-experience', your experience will be published to /experience/my-new-experience.
    </p>
  </div>
  <div
    aria-describedby="body-helper-text"
    className="MuiFormControl-root-3 ExperienceForm-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="body"
    >
      Description
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-multiline-23 MuiInput-underline-21"
    >
      <textarea
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25 MuiInput-inputMultiline-27"
        defaultValue="test"
        disabled={false}
        id="body"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={6}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="body-helper-text"
    >
      Describe your experience so people will know what to expect before they enter your experience.
    </p>
  </div>
  <button
    className="MuiButtonBase-root-50 MuiButton-root-34 MuiButton-raised-40 MuiButton-raisedPrimary-41 ExperienceForm-button-2"
    disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="submit"
  >
    <span
      className="MuiButton-label-35"
    >
      Save
    </span>
    <span
      className="MuiTouchRipple-root-53"
    />
  </button>
</form>
`;
